internal/layer2/merge.go

```diff
@@ -0,0 +1,146 @@
+package layer2
+
+import (
+	"fmt"
+	"strings"
+
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	cloudConfigHeader = "#cloud-config"
+	jinjaHeader      = "## template: jinja"
+)
+
+// CloudConfigMerger handles the merging of cloud configs
+type CloudConfigMerger struct {
+}
+
+// NewCloudConfigMerger creates a new instance of CloudConfigMerger
+func NewCloudConfigMerger() *CloudConfigMerger {
+	return &CloudConfigMerger{}
+}
+
+// configHeaders represents the headers found in a cloud-config file
+type configHeaders struct {
+	hasJinja      bool
+	hasCloudConfig bool
+}
+
+// stripHeaders removes the template and cloud-config headers and returns the remaining content
+func stripHeaders(data string) (string, configHeaders) {
+	headers := configHeaders{}
+	lines := strings.Split(strings.TrimSpace(data), "\n")
+	startIndex := 0
+
+	for i, line := range lines {
+		trimmedLine := strings.TrimSpace(line)
+		switch trimmedLine {
+		case jinjaHeader:
+			headers.hasJinja = true
+			startIndex = i + 1
+		case cloudConfigHeader:
+			headers.hasCloudConfig = true
+			startIndex = i + 1
+		default:
+			if trimmedLine != "" && !strings.HasPrefix(trimmedLine, "#") {
+				return strings.Join(lines[startIndex:], "\n"), headers
+			}
+		}
+	}
+	return "", headers
+}
+
+// deepMerge recursively merges two maps
+func (m *CloudConfigMerger) deepMerge(base, overlay map[string]interface{}) map[string]interface{} {
+	result := make(map[string]interface{})
+
+	// Copy base map
+	for k, v := range base {
+		result[k] = v
+	}
+
+	// Merge overlay
+	for k, v := range overlay {
+		if baseVal, exists := result[k]; exists {
+			// If both values are maps, merge them recursively
+			if baseMap, ok := baseVal.(map[string]interface{}); ok {
+				if overlayMap, ok := v.(map[string]interface{}); ok {
+					result[k] = m.deepMerge(baseMap, overlayMap)
+					continue
+				}
+			}
+
+			// If both values are slices, append them
+			if baseSlice, ok := baseVal.([]interface{}); ok {
+				if overlaySlice, ok := v.([]interface{}); ok {
+					result[k] = append(baseSlice, overlaySlice...)
+					continue
+				}
+			}
+		}
+
+		// For all other cases, overlay value takes precedence
+		result[k] = v
+	}
+
+	return result
+}
+
+// buildHeader constructs the appropriate header based on the input configurations
+func buildHeader(bootstrapHeaders, layer2Headers configHeaders) string {
+	var headers []string
+
+	// If either input has the Jinja header, include it in the output
+	if bootstrapHeaders.hasJinja || layer2Headers.hasJinja {
+		headers = append(headers, jinjaHeader)
+	}
+
+	// Always include the cloud-config header
+	headers = append(headers, cloudConfigHeader)
+
+	return strings.Join(headers, "\n")
+}
+
+// MergeConfigs combines bootstrap data with layer2 config
+func (m *CloudConfigMerger) MergeConfigs(bootstrapData string, layer2UserData string) (string, error) {
```

Comment:
can we make this generic as well ?

```go
CloudConfigMerger.WithBootstrapData(bdata).
WithLayer2Data(l2data).
WithAddlUserDatas([]userdatas). // this can be optional and can be extensible
MergeConfigs()
```